### PR TITLE
[LBSE] Errorous/unnecessary repainting when viewBox is used on <svg> elements

### PR DESCRIPTION
**Pull request by nikolaszimmermann**
#### af30a9bb6eef26b7e5f32b8311ecc0d9ac7df8c1
<pre>
[LBSE] Errorous/unnecessary repainting when viewBox is used on &lt;svg&gt; elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=258351">https://bugs.webkit.org/show_bug.cgi?id=258351</a>

Reviewed by Rob Buis.

The anonymous RenderSVGViewportContainer, that encloses the entire SVG render tree,
caused errorous repaints whenever the outermost &lt;svg&gt; specified a viewBox attribute.

RenderLayerBacking::isSimpleContainerCompositingLayer() returned &apos;false&apos; for the
anyonmous RenderSVGViewportContainer, as it advertized itself as &apos;paintsContents()=true&apos;,
which is incorrect, since it does not paint on its own.

Alter RenderLayer::computeHasVisibleContent() to return false for the anonymous
RenderSVGViewportContainer, similar to the existing condition, that prevents
repaints for layers that have a &apos;RenderSVGHiddenContainer&apos; ancestor in the render
tree hierarchy.

Covered by new tests (graphics layer tree dump will showed &apos;drawsContent 1&apos; before and
&apos;contentsVisible 0&apos; after this change for the anonymous RenderSVGViewportContainer layer).

* LayoutTests/platform/ios/svg/compositing/anonymous-RenderSVGViewportContainer-no-repaints-expected.txt: Added.
  Taken from iOS 16 Simulator EWS build -- scrollbars on iOS influence the GraphicsLayer tree dump, so this needs
  a platform specific result.
* LayoutTests/platform/mac-wk1/svg/compositing/transform-change-repainting-no-viewBox-repaintRects-expected.txt: Added.
* LayoutTests/platform/mac-wk1/svg/compositing/transform-change-repainting-viewBox-repaintRects-expected.txt: Added.
  Legacy WK1 exposes individual repainting bugs -- add an exception instead of skipping to document this.
* LayoutTests/svg/compositing/anonymous-RenderSVGViewportContainer-no-repaints-expected.txt: Added.
* LayoutTests/svg/compositing/anonymous-RenderSVGViewportContainer-no-repaints.html: Added.
* LayoutTests/svg/compositing/transform-change-repainting-no-viewBox-expected.html: Added.
* LayoutTests/svg/compositing/transform-change-repainting-no-viewBox-repaintRects-expected.txt: Added.
* LayoutTests/svg/compositing/transform-change-repainting-no-viewBox-repaintRects.html: Added.
* LayoutTests/svg/compositing/transform-change-repainting-no-viewBox.html: Added.
* LayoutTests/svg/compositing/transform-change-repainting-viewBox-expected.html: Added.
* LayoutTests/svg/compositing/transform-change-repainting-viewBox-repaintRects-expected.txt: Added.
* LayoutTests/svg/compositing/transform-change-repainting-viewBox-repaintRects.html: Added.
* LayoutTests/svg/compositing/transform-change-repainting-viewBox.html: Added.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::computeHasVisibleContent const):

Canonical link: <a href="https://commits.webkit.org/269034@main">https://commits.webkit.org/269034@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/29f443b90c66873bc54d7d91f17915a75d4c0ef2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21323 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21647 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22376 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23185 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19775 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21562 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24933 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21885 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20988 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21546 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21208 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18481 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24037 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18375 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19329 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25658 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19449 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19539 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23499 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20025 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17050 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19334 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/19147 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5115 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23587 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19922 "Built successfully") | | |

### CODE REVIEW

**Comment by nikolaszimmermann** [comment]
Anyone has a suggestion for testing this? I haven't investigated yet, but I bet we have tests for HTML/CSS testing similar things.

I could try the following:
- svg document with viewBox set
- composited <rect> element
- moving <rect> element by e.g. rAF()
- observe/track repaint rects?

File: LayoutTests/svg/compositing/anonymous-RenderSVGViewportContainer-no-repaints.html
Comment (rwlbuis) [inline LayoutTests/svg/compositing/anonymous-RenderSVGViewportContainer-no-repaints.html]: Should this do LayerBasedSVGEngineEnabled=true directive?

File: LayoutTests/svg/compositing/anonymous-RenderSVGViewportContainer-no-repaints.html
Comment (rwlbuis) [inline LayoutTests/svg/compositing/anonymous-RenderSVGViewportContainer-no-repaints.html]: Can I see the effect of this in the expected.txt?

File: LayoutTests/svg/compositing/anonymous-RenderSVGViewportContainer-no-repaints.html
Comment (nikolaszimmermann) [reply LayoutTests/svg/compositing/anonymous-RenderSVGViewportContainer-no-repaints.html]: Good catch!

File: LayoutTests/svg/compositing/anonymous-RenderSVGViewportContainer-no-repaints.html
Comment (nikolaszimmermann) [reply LayoutTests/svg/compositing/anonymous-RenderSVGViewportContainer-no-repaints.html]: Nope, this is a relict from an attempt where I wanted this to be a retest... will remove.

File: LayoutTests/svg/compositing/anonymous-RenderSVGViewportContainer-no-repaints.html
Comment (nikolaszimmermann) [reply LayoutTests/svg/compositing/anonymous-RenderSVGViewportContainer-no-repaints.html]: Was also missing the "!" which I missed initially :-)

**Review by rwlbuis — APPROVED** [review]
LGTM, but "errorous" is "errorous" :)

**Comment by nikolaszimmermann** [comment]
Cannot land this, there are issues in the stress tests, will investigate.

**Comment by nikolaszimmermann** [comment]
This needed rebaselines after PR #18692 landed. Furthermore iOS platform specific results are necessary for one tests, due to scrollbar differences in GraphicsLayer tree dump.

**Comment by nikolaszimmermann** [comment]
Incorporated the mac-wk1 results -- WK1 shows individual repaint issues with LBSE (overpainting). Document this instead of skipping the test with a platform specific result. 

Do we care for WK1 except the bots? AFAIK it only exists on macOS for legacy reasons?

**Pull request merged**
